File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        maxLength = 0
        def dfs(node, direction, length):
            nonlocal maxLength
            if not node:
                return
            maxLength = max(maxLength, length)
            if direction == "left":
                dfs(node.left, "right", length + 1)
                dfs(node.right, "left", 1)
            else:
                dfs(node.right, "left", length + 1)
                dfs(node.left, "right", 1)
        if not root:
            return 0
        
        dfs(root, "left", 0)
        dfs(root, "right", 0)
        return maxLength

File: 1372-longest-zigzag-path-in-a-binary-tree/1372-longest-zigzag-path-in-a-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func dfs(root *TreeNode, isLeft bool, count int) int {
    if root == nil {
        return count
    }

    if isLeft {
        return max(dfs(root.Right, false, count+1), dfs(root.Left, true, 0))
    }

    return max(dfs(root.Left, true, count+1), dfs(root.Right, false, 0))
}

func longestZigZag(root *TreeNode) int {
    return max(dfs(root.Left, true, 0), dfs(root.Right, false, 0))
}

func max(a, b int) int {
    if a > b {
        return a
    }

    return b
}